server/src/future.rs

```diff
@@ -152,92 +146,89 @@ where
 	}
 }
 
-#[derive(Debug)]
-struct MonitorInner {
-	shutdown_requested: AtomicBool,
-	waker: AtomicWaker,
-}
-
-/// Monitor for checking whether the server has been flagged to shut down.
 #[derive(Debug, Clone)]
-pub(crate) struct StopMonitor(Arc<MonitorInner>);
+pub(crate) struct StopHandle(watch::Receiver<()>);
 
-impl Drop for StopMonitor {
-	fn drop(&mut self) {
-		if Arc::strong_count(&self.0) == 1 {
-			self.0.waker.wake();
-		}
-	}
-}
-
-impl StopMonitor {
-	pub(crate) fn new() -> Self {
-		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+impl StopHandle {
+	pub(crate) fn new(rx: watch::Receiver<()>) -> Self {
+		Self(rx)
 	}
 
 	pub(crate) fn shutdown_requested(&self) -> bool {
-		// We expect this method to be polled frequently, skipping an iteration isn't problematic, so relaxed
-		// ordering is optimal.
-		self.0.shutdown_requested.load(Ordering::Relaxed)
+		// if a message has been seen, it means that `stop` has been called.
+		self.0.has_changed().unwrap_or(true)
 	}
 
-	pub(crate) fn handle(&self) -> ServerHandle {
-		ServerHandle(Arc::downgrade(&self.0))
+	pub(crate) async fn shutdown(&mut self) {
+		// Err(_) implies that the `sender` has been dropped;
+		// Ok(_) implies that `stop` has been called,
+		let _ = self.0.changed().await;
 	}
 }
 
-/// Handle that is able to stop the running server or wait for it to finish
-/// its execution.
-#[derive(Debug, Clone)]
-pub struct ServerHandle(Weak<MonitorInner>);
+/// Server handle.
+///
+/// When all server handles has been `dropped` or `stop` has been called
+/// the server will be stopped.
+#[derive(Debug)]
```

Comment:
I have removed the `Clone` but I can add it back.

What's the benefit of having clone here?

Reply:
yeah, I know but I realized that is quite tricky to workaround that if you want to put it in an `Arc` or `Mutex` as `stop` takes ownership so I reverted it back and put it in `Arc` internally and it clone-able again. .